compiler/AST/AggregateType.cpp

```diff
@@ -2528,3 +2532,87 @@ Symbol* AggregateType::getSubstitution(const char* name) {
 
   return retval;
 }
+
+bool AggregateType::isRawClass() const {
+  return aggregateTag == AGGREGATE_CLASS && classKind == CLASS_RAW;
+}
+
+bool AggregateType::isBorrowedClass() const {
+  return aggregateTag == AGGREGATE_CLASS && classKind == CLASS_BORROW;
+}
+
+bool AggregateType::isCanonicalClass() const {
+  return (this->defaultClass == NULL || this->defaultClass == this);
+}
+
+
+AggregateType* AggregateType::getRawClass() {
+  if (aggregateTag == AGGREGATE_CLASS) {
+    if (isRawClass())
+      return this;
+
+    generateRawBorrowClassTypes();
+
+    for (AggregateType* t = this->defaultClass;
+         t != NULL;
+         t = t->nextAssociatedClass) {
+      if (t->isRawClass())
+        return t;
+    }
+  }
+  return NULL;
+}
+AggregateType* AggregateType::getBorrowedClass() {
+  if (aggregateTag == AGGREGATE_CLASS) {
+    if (isBorrowedClass())
+      return this;
+
+    generateRawBorrowClassTypes();
+
+    for (AggregateType* t = this->defaultClass;
+         t != NULL;
+         t = t->nextAssociatedClass) {
+      if (t->isBorrowedClass())
+        return t;
+    }
+  }
+  return NULL;
+}
+
+AggregateType* AggregateType::getCanonicalClass() {
+  if (aggregateTag == AGGREGATE_CLASS) {
+    generateRawBorrowClassTypes();
+    INT_ASSERT(this->defaultClass);
+    return this->defaultClass;
+  }
+
+  return NULL;
+}
+
+
+void AggregateType::generateRawBorrowClassTypes() {
+  AggregateType* at = this;
+  if (aggregateTag == AGGREGATE_CLASS && at->defaultClass == NULL) {
+    SET_LINENO(at->symbol->defPoint);
+    // Generate raw and owned class types
+    AggregateType* raw = new AggregateType(AGGREGATE_CLASS);
+
+    at->classKind = CLASS_BORROW;
+    raw->classKind = CLASS_RAW;
+
+    at->defaultClass = at;
+    raw->defaultClass = at;
+    at->nextAssociatedClass = raw;
+
+
+    TypeSymbol* tsRaw = new TypeSymbol(astr("raw ", at->symbol->name), raw);
```

Comment:
Is changing the keyword to unmanaged sufficient for this? Note that the name of this type symbol is visible to users (e.g. in compiler errors, type to string output). For that reason I think the space is more natural since it matches the syntax we are trying to support.